numpy/core/arrayprint.py

```diff
@@ -496,6 +497,16 @@ def _array2string(a, options, separator=' ', prefix=""):
     return lst
 
 
+def _array2string_dispatcher(
+        a, max_line_width=None, precision=None,
+        suppress_small=None, separator=None, prefix=None,
+        style=None, formatter=None, threshold=None,
+        edgeitems=None, sign=None, floatmode=None, suffix=None,
+        **kwarg):
+    return (a,)
+
+
+@array_function_dispatch(_array2string_dispatcher)
```

Comment:
Could the dispatcher be written
```
def default1_dispatcher(*args, **kwargs):
    return args[:1]
```
Then maybe a few of those could be put into overrides, imported, and reused. Of course there may be cases for a different dispatcher in special cases.

Reply:
This is a good idea, but even though most numpy functions have a meaningless first argument name, technically you can still use them as a keyword argument, e.g., `numpy.sum(a=array)`. So we need to have an exactly matching function signature, including the same argument names.

Reply:
At least for python 3.3+, we could use inspect.signature and set the dispatcher.__signature__ as in this [stack overflow answer](https://stackoverflow.com/questions/1409295/set-function-signature-in-python), or is that too magical?

Reply:
I wonder if signature rewriting would let us restore the original traceback when the wrong arguments are supplied. With the current version of NumPy:
```
In [2]: np.argmax(x=[1, 2, 3])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-d25c18244167> in <module>()
----> 1 np.argmax(x=[1, 2, 3])

TypeError: argmax() got an unexpected keyword argument 'x'
```
but after this PR:
```
In [2]: np.argmax(x=[1, 2, 3])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-d25c18244167> in <module>()
----> 1 np.argmax(x=[1, 2, 3])

~/dev/numpy/numpy/core/overrides.py in public_api(*args, **kwargs)
    147         @functools.wraps(implementation)
    148         def public_api(*args, **kwargs):
--> 149             relevant_args = dispatcher(*args, **kwargs)
    150             return array_function_implementation_or_override(
    151                 implementation, public_api, relevant_args, args, kwargs)

TypeError: _argmax_dispatcher() got an unexpected keyword argument 'x'
```

Reply:
Replacing `public_api.__signature__` doesn't effect the error message, but it looks like this is something that the [decorator library](https://pypi.org/project/decorator/) would fix. So maybe that's worth doing after all...